stan/math/prim/fun/as_value_column_array_or_scalar.hpp

```diff
@@ -0,0 +1,72 @@
+#ifndef STAN_MATH_PRIM_FUN_AS_VALUE_COLUMN_ARRAY_OR_SCALAR
+#define STAN_MATH_PRIM_FUN_AS_VALUE_COLUMN_ARRAY_OR_SCALAR
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Extract values from var types and containers and return a column vector or a
+ * scalar. For arithmetic scalar inputs that is an identity function.
+ *
+ * @tparam T Type of scalar element.
+ * @param a Specified scalar.
+ * @return the scalar.
+ */
+template <typename T, require_stan_scalar_t<T>* = nullptr>
+inline auto as_value_column_array_or_scalar(const T& a) {
+  return value_of(a);
+}
+
+/**
+ * Extract values from var types and containers and return a column vector or a
+ * scalar. For arithmetic column vector inputs this is an identity function.
+ *
+ * @tparam T Type of scalar element.
+ * @param a Specified vector.
+ * @return Same vector.
+ */
+template <typename T, require_col_vector_t<T>* = nullptr>
+inline auto as_value_column_array_or_scalar(T&& a) {
+  return value_of(std::forward<T>(a)).array();
```

Comment:
I guess const refs aren't gonna work if we want the result here to be assignable?

Reply:
Yes. We this function needs to accept a forwarding reference to handle both lvalues and rvalues. The forward call, however, is not necessary, as `.array()` does not handle rvalues anyway. Instead you need a holder.

Reply:
So for this function we're assuming we might pass it values that have not been held, right?

To make a holder make sense here, I think we either need to assume `a` is not held, or `.array()` allocates memory on the heap.

`value_of` will hold its input if necessary.

Reply:
What do you mean by "being held"? Also afaik `.array()` does not allocate anything on heap.

Reply:
Because this function takes `T&&`, we have to worry about the input being an rvalue (it might not be held).

For comparison, [this](https://github.com/stan-dev/math/blob/develop/stan/math/prim/fun/transpose.hpp#L18) function takes `const &` which can be an rvalue, and it does a `.transpose()` which is a non-heap-allocating-Eigen-view (like `.array()`), and we don't need a holder there.

The reason we decided we didn't need a holder there was we assumed all inputs that needed held have been held right?

Reply:
That is not actually how it works. If an argument is a rvalue it will go out of scope regardless of whether the function takes that argument as a `const &` or `&&`. The difference is we are assuming that no one will do stuff like `auto x = transpose(rvalue)`. As I wrote in another comment: maybe we should extend this assumption to internal function (and enforce it by requires).

So by `an argument not being held` you mean the argument is a rvalue?

Reply:
> So by an argument not being held you mean the argument is a rvalue?

I think I mean, safe to use without worrying about holders.

By my reading holder is more associated with how Eigen defines the memory in expressions, not really by rvalue-ness. There are things that are rvalues that will not cause problems (for instance, expression rvalues).

> The difference is we are assuming that no one will do stuff like auto x = transpose(rvalue).

I read the holder docs again, and the reason we do not need a holder for that is because the temporary we created is `m.transpose()` and this is a lightweight Eigen expression and can be returned directly.

In this situation, `value_of` might return a heavy object, but because `value_of` knows about holders, it will create one if necessary to hold the memory temporary.

Effectively we have:

```
holder<S> held_values = value_of(std::forward<T>(a));
return held_values.transpose();
```

So if `held_values` ever gets deleted, the memory in `value_of` gets deleted and everything goes up in smoke.

My question now is, since `held_values` will already be a holder, do we need to hold that? I'd expect that the way it is written (as a temporary, and not like an lvalue), it will just move itself along and we won't need another holder to wrap it.

Reply:
> I think I mean, safe to use without worrying about holders.

That depends a lot about assumptions. In most cases I can understand that as is a lvalue, but there are exceptions. Can you be more specific?

> By my reading holder is more associated with how Eigen defines the memory in expressions, not really by rvalue-ness.

Yeah, that in turn is tightly associated with the fact Eigen in most cases does not handle rvalues as they are intended to be handled.

> There are things that are rvalues that will not cause problems

That depends a lot on assumptions. Any rvalue function argument that is referenced by return value is a potential problem.

> I read the holder docs again, and the reason we do not need a holder for that is because the temporary we created is m.transpose() and this is a lightweight Eigen expression and can be returned directly.

Yes, but the rvalue the `transpose` function is called with can go out of scope before that transposition is evaluated.

>My question now is, since held_values will already be a holder, do we need to hold that? I'd expect that the way it is written (as a temporary, and not like an lvalue), it will just move itself along and we won't need another holder to wrap it.

Yes we do. Memory in holder is not permanent (that would be a memory leak). It exists only as long as holder. In C++ nothing just moves itself along on its own. The programmer has to implement such stuff.

Reply:
> In most cases I can understand that as is a lvalue, but there are exceptions.

This is why I think lvalue/rvalue are the wrong way to think about when to use a holder or not.

How about:

When returning an Eigen expression, if the expression depends on anything other than arguments to the function that allocates memory on the heap, we need a holder.

> Memory in holder is not permanent (that would be a memory leak).

Mm, so a holder is good for only one level of return and then destructs. So is making a new holder from the old holder cheap? That is, the variables stored by the old holder are moved to the new one, and nothing is copied or destructed?

Reply:
> When returning an Eigen expression, if the expression depends on anything other than arguments to the function that allocates memory on the heap, we need a holder.

That is just one case. We are also using holder for arguments of internal functions (such as ones added in this PR) that are likely to have their return value assigned to an auto variable.

> So is making a new holder from the old holder cheap?

Making holder is in no case very expensive. In most cases it the worst thing is a memory allocation.

> That is, the variables stored by the old holder are moved to the new one, and nothing is copied or destructed?

Not at the moment, but I think we can do that.

Reply:
> We are also using holder for arguments of internal functions (such as ones added in this PR) that are likely to have their return value assigned to an auto variable.

So this is:

```
auto x = myfunc(y);
```

`auto x` is copy-constructed from the temporary in `myfunc(y)`, and then the temporary `myfunc(y)` is destructed. So in this case even if `myfunc(y)` returned a holder, `x` is not safe to use because the holder got destructed?

Reply:
Nope. `x` is either move-constructed or the compiler does copy elision. So it is safe.

Reply:
Alright so assigning to an auto variable isn't a problem, and so I am back to just the one rule for holders:

> When returning an Eigen expression, if the expression depends on anything other than arguments to the function that allocates memory on the heap, we need a holder.

More exceptions more exceptions and I will add more rules

Reply:
Well if we don't have any exceptions to this rule, then this is the rule for when to use `holder` or not.

Reply:
As I explained earlier we do have exceptions:

> That is just one case. We are also using holder for arguments of internal functions (such as ones added in this PR) that are likely to have their return value assigned to an auto variable.

Reply:
Yes but then I made a function which returned a holder return it to an auto variable and it wasn't a problem because of move constructor/copy elision optimizations (this [one](https://github.com/stan-dev/math/pull/2278#discussion_r551931867)).

So can you be more specific with the example of when auto causes a problem?

Reply:
If the function is called with a rvalue argument that owns dynamically allocated memory. So this will likely (but necessarily) blow up (segfault / wrong result), if func does not handle rvalues (for example via using Holder):
```
MatrixXd x = ...
auto y = func(eval(x.transpose()));
MatrixXd z = y;
```
If more code is between last two lines the probability of errors increases.

Reply:
```
func(eval(x.transpose()))
```

In this case the function [here](https://github.com/stan-dev/math/blob/develop/stan/math/prim/fun/transpose.hpp#L18) needs a holder too right?

And if that code doesn't need a holder, then we must be making assumptions about our function inputs (that the situation you outline above doesn't arise, or something).